contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,368 @@
+    /// @param _requestId request id to claim
+    /// @dev will use `findClaimHintUnbounded()` to find a hint, what can lead to OOG
+    /// Prefer `claimWithdrawal(uint256 _requestId, uint256 _hint)` to save gas
+    function claimWithdrawal(uint256 _requestId) external {
+        _claimWithdrawalTo(_requestId, findClaimHintUnbounded(_requestId), msg.sender);
+        _emitTransfer(msg.sender, address(0), _requestId);
+    }
+
+    /// @notice Finds the list of hints for the given `_requestIds` searching among the checkpoints with indices
+    ///  in the range  `[_firstIndex, _lastIndex]`
+    /// @param _requestIds ids of the requests sorted in the ascending order to get hints for
+    /// @param _firstIndex left boundary of the search range
+    /// @param _lastIndex right boundary of the search range
+    /// @return hintIds the hints for `claimWithdrawal` to find the checkpoint for the passed request ids
+    function findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex)
+        public
+        view
+        returns (uint256[] memory hintIds)
+    {
+        hintIds = new uint256[](_requestIds.length);
+        uint256 prevRequestId = 0;
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            if (_requestIds[i] < prevRequestId) revert RequestIdsNotSorted();
+            hintIds[i] = findClaimHint(_requestIds[i], _firstIndex, _lastIndex);
+            _firstIndex = hintIds[i];
+            prevRequestId = _requestIds[i];
+        }
+    }
+
+    /// @notice Finds the list of hints for the given `_requestIds` searching among the checkpoints with indices
+    ///  in the range `[1, lastCheckpointIndex]`
+    /// @dev WARNING! OOG is possible if used onchain.
+    ///  See `findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex)` for onchain use
+    /// @param _requestIds ids of the requests sorted in the ascending order to get hints for
+    function findClaimHintsUnbounded(uint256[] calldata _requestIds) public view returns (uint256[] memory hintIds) {
+        return findClaimHints(_requestIds, 1, getLastCheckpointIndex());
+    }
+
+    /// @notice Finalize requests from last finalized one up to `_lastRequestIdToFinalize`
+    /// @dev ether to finalize all the requests should be calculated using `finalizationBatch()` and sent along
+    ///
+    /// @param _nextFinalizedRequestId request index in the queue that will be last finalized request in a batch
+    function finalize(uint256 _nextFinalizedRequestId) external payable whenResumed onlyRole(FINALIZE_ROLE) {
+        _finalize(_nextFinalizedRequestId, msg.value.toUint128());
+    }
+
+    /// @notice Update bunker mode state
+    /// @dev should be called by oracle
+    ///
+    /// @param _isBunkerModeNow oracle report
+    /// @param _bunkerModeSinceTimestamp timestamp of start of the bunker mode
+    function updateBunkerMode(bool _isBunkerModeNow, uint256 _bunkerModeSinceTimestamp)
```

Comment:
`_bunkerModeSinceTimestamp` -> `_previousReportTimestamp`

Reply:
`enableBunker(uint256  retroactiveStartTimestamp)`
`disableBunker()`

Reply:
Looks like too much logic is going to the oracle if we divide the method. And it's alien for oracle contract as well. So , I left it as is.